ndonnx/extensions.py

```diff
@@ -0,0 +1,437 @@
+
+from __future__ import annotations
+
+from collections.abc import Mapping, Sequence
+from typing import Literal, TypeAlias, TypeVar, get_args
+from warnings import warn
+
+import numpy as np
+from typing_extensions import TypeIs
+
+import ndonnx as ndx
+
+from . import _typed_array as tydx
+from ._typed_array.masked_onnx import TyMaArray
+
+Scalar = TypeVar("Scalar", int, float, str)
+
+Key: TypeAlias = Scalar
+Value = TypeVar("Value", int, float, str)
+
+
+def shape(x: ndx.Array, /) -> ndx.Array:
+    """Returns shape of an array.
+
+    Parameters
+    ----------
+    x: Array
+        Array to get shape of
+
+    Returns
+    -------
+    out: Array
+        Array of shape
+    """
+    warn(
+        "'ndonnx.shape' is deprecated in favor of 'ndonnx.Array.dynamic_shape'",
+        DeprecationWarning,
+    )
+    return x.dynamic_shape
+
+
+def isin(x: ndx.Array, /, items: Sequence[Scalar]) -> ndx.Array:
+    """Return true where the input ``Array`` contains an element in ``items``.
+
+    ``NaN`` values do **not** compare equal.
+
+    Parameters
+    ----------
+    x: Array
+        The input Array to check for the presence of items.
+    items: Sequence[Scalar]
+        Scalar items to check for in the input Array.
+
+    Returns
+    -------
+    out: Array
+        Array of booleans indicating whether each element of ``x`` is in ``items``.
+    """
+    return ndx.Array._from_tyarray(x._tyarray.isin(items))
+
+
+# TODO: Bad naming: It is obvious from the type hints that this mapping is "static"
+def static_map(
+    x: ndx.Array,
+    /,
+    mapping: Mapping[Key, Value],
+    default: Value | None = None,
+) -> ndx.Array:
+    """Map values in ``x`` based on the static ``mapping``.
+
+    Parameters
+    ----------
+    x: Array
+        The Array whose values will be mapped.
+    mapping: Mapping[Key, Value]
+        A mapping from keys to values. The keys must be of the same type as the values in ``x``.
+    default: Value, optional
+        The default value to use when a key is not found in the mapping.
+
+    Returns
+    -------
+    out: Array
+        A new Array with the values mapped according to the mapping.
+    """
+    values = np.array(list(mapping.values()))
+
+    if values.dtype.kind in ("O", "U"):
+        values = values.astype(str)
+
+    if default is None:
+        # Default values do not follow the ONNX standard (which is ok!)
+        if np.issubdtype(values.dtype, np.floating):
+            default = 0.0  # type: ignore
+        elif np.issubdtype(values.dtype, np.integer):
+            default = 0  # type: ignore
+        elif values.dtype.kind == "U":
+            default = "MISSING"  # type: ignore
+        elif values.dtype == bool:
+            # to be in line with the numerical 0
+            default = False  # type: ignore
+    if default is None:
+        raise TypeError(
+            f"failed to infer default value for mapping values of data type `{values.dtype}`"
+        )
```

Comment:
Is there a reason this isn't happening in the typed array implementation?

Reply:
I added the defaults to the docstring now, which I think gives this code its logical place here. The typed array member functions do not have an optional `default`.